Parsing of floating-point number values in attributes should allow leading ascii whitespace, plus sign, and trailing junk

sideshowbarker opened the pull request:
#### 806925c1274e707952092fa37779fd4884587985
<pre>
Parsing of floating-point number values in attributes should allow leading ascii whitespace, plus sign, and trailing junk
<a href="https://bugs.webkit.org/show_bug.cgi?id=268542">https://bugs.webkit.org/show_bug.cgi?id=268542</a>

Reviewed by Darin Adler.

For any attribute whose value the HTML spec defines as a “floating-point
number” (WebIDL “double”), this change makes WebKit apply the spec
requirements in the “rules for parsing floating-point number values” at
<a href="https://html.spec.whatwg.org/#rules-for-parsing-floating-point-number-values">https://html.spec.whatwg.org/#rules-for-parsing-floating-point-number-values</a>
when it should — and thus correctly skip/ignore:

  - any leading ASCII whitespace
  - any leading plus signs
  - any trailing “junk” characters that aren’t ASCII digits

…while also ensuring the U+000B LINE TABULATION (vertical tab) character is
not skipped/ignored — because although Unicode defines U+000B as whitespace,
per <a href="https://infra.spec.whatwg.org/#ascii-whitespace">https://infra.spec.whatwg.org/#ascii-whitespace</a>, U+000B is not among the
set of “ASCII whitespace” characters the HTML spec allows ignoring/skipping
when applying the “rules for parsing floating-point number values” algorithm.

Otherwise, without this change, WebKit doesn’t apply the relevant spec
requirements when it should, and doesn’t correctly ignore any leading ASCII
whitespace or plus signs, nor ignore any trailing junk — and would also
incorrectly include U+000B among the set of ignorable whitespace characters.

Note: The spec states that the “valid floating-point number” algorithm at
<a href="https://html.spec.whatwg.org/#valid-floating-point-number">https://html.spec.whatwg.org/#valid-floating-point-number</a> must be applied
when parsing certain attribute values, and that algorithm requires rejecting
values with leading ASCII whitespace or leading plus signs and trailing junk.
But for other attribute values, the spec states that the algorithm at
<a href="https://html.spec.whatwg.org/#rules-for-parsing-floating-point-number-values">https://html.spec.whatwg.org/#rules-for-parsing-floating-point-number-values</a>
must be applied instead — and that requires ignoring/skipping any leading
ASCII whitespace or plus signs, and ignoring/skipping any trailing junk.

* LayoutTests/fast/dom/HTMLMeterElement/set-meter-properties-expected.txt:
* LayoutTests/fast/dom/HTMLMeterElement/set-meter-properties.html:
* LayoutTests/fast/dom/HTMLProgressElement/set-progress-properties-expected.txt:
* LayoutTests/fast/dom/HTMLProgressElement/set-progress-properties.html:
* LayoutTests/imported/w3c/web-platform-tests/html/dom/reflection-forms-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/html/dom/reflection-misc-expected.txt:
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/html/dom/reflection-misc-expected.txt: Added.
* LayoutTests/platform/ios/imported/w3c/web-platform-tests/html/dom/reflection-forms-expected.txt: Removed.
* LayoutTests/platform/mac-wk1/imported/w3c/web-platform-tests/html/dom/reflection-forms-expected.txt: Added.
* LayoutTests/platform/mac-wk1/imported/w3c/web-platform-tests/html/dom/reflection-misc-expected.txt: Added.
* LayoutTests/platform/mac-wk2/imported/w3c/web-platform-tests/html/dom/reflection-forms-expected.txt: Removed.
* LayoutTests/platform/mac-wk2/imported/w3c/web-platform-tests/html/dom/reflection-misc-expected.txt: Removed.
* LayoutTests/platform/wpe/imported/w3c/web-platform-tests/html/dom/reflection-embedded-expected.txt: Removed.
* LayoutTests/platform/wpe/imported/w3c/web-platform-tests/html/dom/reflection-forms-expected.txt:
* Source/WebCore/html/HTMLMeterElement.cpp:
(WebCore::HTMLMeterElement::min const):
(WebCore::HTMLMeterElement::max const):
(WebCore::HTMLMeterElement::value const):
(WebCore::HTMLMeterElement::low const):
(WebCore::HTMLMeterElement::high const):
(WebCore::HTMLMeterElement::optimum const):
* Source/WebCore/html/HTMLProgressElement.cpp:
(WebCore::HTMLProgressElement::value const):
(WebCore::HTMLProgressElement::max const):
* Source/WebCore/html/parser/HTMLParserIdioms.cpp:
(WebCore::parseHTMLFloatingPointNumberValueInternal):
(WebCore::parseHTMLFloatingPointNumberValue):
* Source/WebCore/html/parser/HTMLParserIdioms.h:

Canonical link: <a href="https://commits.webkit.org/278541@main">https://commits.webkit.org/278541@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/61ecd74b5a40eee952fbd6a6ad9cf02744481386

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/50809 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/30106 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/3127 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/54068 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/1500 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/53112 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/36379 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/1150 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/41392 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/52908 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/27755 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/43772 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/22520 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/25100 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/1028 "Passed tests") | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/9250 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/47081 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/1090 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/55662 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/25911 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/970 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/48802 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/27168 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/43851 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/47878 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/11148 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/28036 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/26900 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

annevk commented on Source/WebCore/html/parser/HTMLParserIdioms.cpp:
```diff
@@ -92,10 +92,15 @@ double parseToDoubleForNumberType(StringView string, double fallbackValue)
+    for (uint i = 0; i < string.length(); i++) {
+        if (string[i] == '\v')
+            return fallbackValue;
+    }
```
I think instead we should change how we parse doubles or make the type of whitespace configurable at that level. Otherwise we'll continue to run into this as we add new callers.

It does seem reasonable to me to leave the `\v` case out for now and leave those tests as a failure. Then this would still be a good WPT progression, but without introducing weirdness into the code base.

tbroyer replied:
IIUC, that check would also incorrectly detect a `\v` in the "trailing junk" (e.g. `"5%\v"` or `"7\v"`) and reject the value altogether instead of parsing the number out of it (here, `5` or `7` respectively)

sideshowbarker replied:
> IIUC, that check would also incorrectly detect a `\v` in the "trailing junk" (e.g. `"5%\v"` or `"7\v"`) and reject the value altogether instead of parsing the number out of it (here, `5` or `7` respectively)

Indeed. From trying this, it’s clear that attempting to special-case `\v` at this point in the code not ideal anyway (even when updating the existing patch here to correctly handle the trailing-junk case), and that properly making WebKit conform to the spec on this will require changing WebKit’s `String::toDouble()` code (and maybe making changes to at least some of the call sites for it)

tbroyer commented on LayoutTests/imported/w3c/web-platform-tests/html/dom/reflection.js:
```diff
@@ -522,24 +600,69 @@ ReflectionTests.typeMap = {
-        "idlDomExpected": ["-10000000000", "-1", "0", "0", "1", "10000000000"],
-        "idlIdlExpected": [ -10000000000,   -1,  -0,   0,   1,   10000000000]
+        "idlDomExpected": function (val) {
+            return Number(val).toFixed();
```
Fwiw, I updated the WPT pull request, as this was actually wrong (won't make a difference for the 5 values being tested though).

tbroyer commented on Source/WebCore/html/parser/HTMLParserIdioms.cpp:
```diff
@@ -107,7 +102,7 @@ double parseToDoubleForNumberType(StringView string, double fallbackValue)
 
     bool valid = false;
-    double value = string.toDouble(valid);
+    double value = string.toDouble(valid, true);
```
Be careful of the side effect on HTMLInputElement in the Number state, as pointed out by https://html.spec.whatwg.org/multipage/common-microsyntaxes.html#floating-point-numbers:valid-floating-point-number-3
IIUC, that one will use that parseToDoubleForNumberType function and expects it to reject invalid values, (including those with leading whitespace, leading plus sign, and/or trailing junk)
https://github.com/WebKit/WebKit/blob/29eeaffe3de46d4c072decbcacb10f99ffffb925/Source/WebCore/html/NumberInputType.cpp#L263-L268
But they should be caught by WPT html/semantics/forms/the-input-element/number.html if I'm reading correctly.

sideshowbarker replied:
Thanks — I changed that `parseToDoubleForNumberType(proposedValue)` call in `NumberInputType::sanitizeValue` to `parseValidHTMLFloatingPointNumber(proposedValue).value()`

But I can’t yet get test results for that `html/semantics/forms/the-input-element/number.html` WPT test, because when I run a build from this PR branch against it, the build crashes — and I’ve not yet been able to get a stack trace from the crash, so it’s going to take me some time to debug it. (The EWS results here show that it’s causing crashes with a couple dozen other test files as well.)

tbroyer reviewed (commented):
LGTM, but my C++ is ~20 year-old rusty and I'm not familiar enough with the WebKit codebase (and only browsed it a bit through the GitHub web UI) to tell whether there could be some impacts outside those files and/or if there could be a better/alternative approach. At least the modified LayoutTests/dast/dom look good to me.

My gut feeling is that HTMLParseIdioms is linked to the HTML spec, so isHTMLDouble might not be the best name here.
Apparently, parseToDoubleForNumberType is also used in HTMLInputElement and RangeInputType, which are defined by the spec to require a "valid floating-point number", so maybe rename isHTMLDouble to requiresValid (and invert the behavior, changing the call sites in HTMLInputElement and RangeInputType and leaving HTMLProgressElement and HTMLMeterElement as-is to get the new behavior), or possibly rename to lenient?
I'll leave that discussion to the WebKit maintainers though.

darinadler commented on Source/WebCore/html/parser/HTMLParserIdioms.cpp:
```diff
@@ -254,6 +254,44 @@ std::optional<double> parseValidHTMLFloatingPointNumber(StringView input)
+        ++leadingSpacesLength;
+
+    position += leadingSpacesLength;
+    if (*position != '+' && *position != '-' && *position != '.' && !isASCIIDigit(*position))
```
This needs to check if `leadingSpacesLength` is `< length` before dereferencing `position`.

darinadler commented on Source/WebCore/html/parser/HTMLParserIdioms.cpp:
```diff
@@ -254,6 +254,44 @@ std::optional<double> parseValidHTMLFloatingPointNumber(StringView input)
+    double number = parseDouble(position, length - leadingSpacesLength, parsedLength);
+
+    // Numbers are considered finite IEEE 754 Double-precision floating point values.
+    ASSERT(-std::numeric_limits<double>::max() <= number || number < std::numeric_limits<double>::max());
```
Not sure why this assertion is needed. It’s OK, but doesn’t seem all that helpful.

darinadler commented on Source/WebCore/html/parser/HTMLParserIdioms.cpp:
```diff
@@ -254,6 +254,44 @@ std::optional<double> parseValidHTMLFloatingPointNumber(StringView input)
 }
 
+template <typename CharacterType>
+static double parseHTMLFloatingPointNumberValueInternal(const CharacterType* position, size_t length, double fallbackValue)
```
We should use `std::span` here instead of a separate `position` and `length`.

darinadler commented on Source/WebCore/html/parser/HTMLParserIdioms.cpp:
```diff
@@ -254,6 +254,44 @@ std::optional<double> parseValidHTMLFloatingPointNumber(StringView input)
+template <typename CharacterType>
+static double parseHTMLFloatingPointNumberValueInternal(const CharacterType* position, size_t length, double fallbackValue)
+{
+    ASSERT(length > 0);
```
This isn’t really needed. Once we add the check that there are characters left after leading spaces, we’ll automatically handle the zero length case correctly.

darinadler commented on Source/WebCore/html/parser/HTMLParserIdioms.cpp:
```diff
@@ -254,6 +254,44 @@ std::optional<double> parseValidHTMLFloatingPointNumber(StringView input)
+double parseHTMLFloatingPointNumberValue(StringView input, double fallbackValue)
+{
+    if (input.isEmpty())
+        return fallbackValue;
```
This probably isn’t needed if you deal with the issues above.

sideshowbarker replied:
```cpp
    if (input.isEmpty())
        return fallbackValue;
```
> This probably isn’t needed if you deal with the issues above.

If we drop that from the code, we get failures for these tests:

https://github.com/WebKit/WebKit/blob/8639b7056d69938550ad662317bd8451b6765abf/LayoutTests/fast/dom/HTMLMeterElement/set-meter-properties.html#L12-L18

…because those need that `fallbackValue` in that _“before properties were set”_ case.

But maybe the failures are due to me not having sufficiently/correctly implemented all the code changes from your review comments?

darinadler replied:
I think if you correctly handle the end of the string case, then `parseHTMLFloatingPointNumberValueInternal` will return `fallbackValue`.

darinadler commented on Source/WebCore/html/parser/HTMLParserIdioms.cpp:
```diff
@@ -254,6 +254,44 @@ std::optional<double> parseValidHTMLFloatingPointNumber(StringView input)
+    if (*position != '+' && *position != '-' && *position != '.' && !isASCIIDigit(*position))
+        return fallbackValue;
+
+    size_t parsedLength = 0;
```
Not sure why we initialize this to 0. Doesn’t seem to be needed.

darinadler reviewed (changes requested):
Need to deal with the buffer overrun. Otherwise this looks great.

darinadler commented on Source/WebCore/html/parser/HTMLParserIdioms.cpp:
```diff
@@ -241,6 +241,37 @@ std::optional<double> parseValidHTMLFloatingPointNumber(StringView input)
+        ++leadingSpacesLength;
+
+    position += leadingSpacesLength;
+    if (leadingSpacesLength < length && *position != '+' && *position != '-' && *position != '.' && !isASCIIDigit(*position))
```
I think we want `if (leadingSpaces == length || (*position ... ))` here instead. If we are out of characters we should return `fallbackValue`. The test cases to see if this is handled correctly would be the empty string (if you remove the empty string check in `parseHTMLFloatingPointNumberValue`), and also strings consisting entirely of ASCII whitespace, in cases where the fallback value is not zero.

sideshowbarker replied:
> I think we want `if (leadingSpaces == length || (*position ... ))` here instead.

aha yes — I see now

> If we are out of characters we should return `fallbackValue`. The test cases to see if this is handled correctly would be the empty string (if you remove the empty string check in `parseHTMLFloatingPointNumberValue`), and also strings consisting entirely of ASCII whitespace, in cases where the fallback value is not zero.

Thanks, yep — tried it now locally, and confirmed

sideshowbarker committed:
Parsing of floating-point number values in attributes should allow leading ascii whitespace, plus sign, and trailing junk
https://bugs.webkit.org/show_bug.cgi?id=268542

Reviewed by Darin Adler.

For any attribute whose value the HTML spec defines as a “floating-point
number” (WebIDL “double”), this change makes WebKit apply the spec
requirements in the “rules for parsing floating-point number values” at
https://html.spec.whatwg.org/#rules-for-parsing-floating-point-number-values
when it should — and thus correctly skip/ignore:

  - any leading ASCII whitespace
  - any leading plus signs
  - any trailing “junk” characters that aren’t ASCII digits

…while also ensuring the U+000B LINE TABULATION (vertical tab) character is
not skipped/ignored — because although Unicode defines U+000B as whitespace,
per https://infra.spec.whatwg.org/#ascii-whitespace, U+000B is not among the
set of “ASCII whitespace” characters the HTML spec allows ignoring/skipping
when applying the “rules for parsing floating-point number values” algorithm.

Otherwise, without this change, WebKit doesn’t apply the relevant spec
requirements when it should, and doesn’t correctly ignore any leading ASCII
whitespace or plus signs, nor ignore any trailing junk — and would also
incorrectly include U+000B among the set of ignorable whitespace characters.

Note: The spec states that the “valid floating-point number” algorithm at
https://html.spec.whatwg.org/#valid-floating-point-number must be applied
when parsing certain attribute values, and that algorithm requires rejecting
values with leading ASCII whitespace or leading plus signs and trailing junk.
But for other attribute values, the spec states that the algorithm at
https://html.spec.whatwg.org/#rules-for-parsing-floating-point-number-values
must be applied instead — and that requires ignoring/skipping any leading
ASCII whitespace or plus signs, and ignoring/skipping any trailing junk.

* LayoutTests/fast/dom/HTMLMeterElement/set-meter-properties-expected.txt:
* LayoutTests/fast/dom/HTMLMeterElement/set-meter-properties.html:
* LayoutTests/fast/dom/HTMLProgressElement/set-progress-properties-expected.txt:
* LayoutTests/fast/dom/HTMLProgressElement/set-progress-properties.html:
* LayoutTests/imported/w3c/web-platform-tests/html/dom/reflection-forms-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/html/dom/reflection-misc-expected.txt:
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/html/dom/reflection-misc-expected.txt: Added.
* LayoutTests/platform/ios/imported/w3c/web-platform-tests/html/dom/reflection-forms-expected.txt: Removed.
* LayoutTests/platform/mac-wk1/imported/w3c/web-platform-tests/html/dom/reflection-forms-expected.txt: Added.
* LayoutTests/platform/mac-wk1/imported/w3c/web-platform-tests/html/dom/reflection-misc-expected.txt: Added.
* LayoutTests/platform/mac-wk2/imported/w3c/web-platform-tests/html/dom/reflection-forms-expected.txt: Removed.
* LayoutTests/platform/mac-wk2/imported/w3c/web-platform-tests/html/dom/reflection-misc-expected.txt: Removed.
* LayoutTests/platform/wpe/imported/w3c/web-platform-tests/html/dom/reflection-embedded-expected.txt: Removed.
* LayoutTests/platform/wpe/imported/w3c/web-platform-tests/html/dom/reflection-forms-expected.txt:
* Source/WebCore/html/HTMLMeterElement.cpp:
(WebCore::HTMLMeterElement::min const):
(WebCore::HTMLMeterElement::max const):
(WebCore::HTMLMeterElement::value const):
(WebCore::HTMLMeterElement::low const):
(WebCore::HTMLMeterElement::high const):
(WebCore::HTMLMeterElement::optimum const):
* Source/WebCore/html/HTMLProgressElement.cpp:
(WebCore::HTMLProgressElement::value const):
(WebCore::HTMLProgressElement::max const):
* Source/WebCore/html/parser/HTMLParserIdioms.cpp:
(WebCore::parseHTMLFloatingPointNumberValueInternal):
(WebCore::parseHTMLFloatingPointNumberValue):
* Source/WebCore/html/parser/HTMLParserIdioms.h:

Canonical link: https://commits.webkit.org/278541@main
